use wintun and windivert for windows platform bits

majestrate opened the pull request:
fixes #710 
fixes #1887
fixes #1981 
fixes #1944
fixes #1953 
fixes #1927
fixes #1939

Commit:
refactor dns subsystem

we want to be able to have multiple locally bound dns sockets in lokinet so
i restructured most of the dns subsystem in order to make this easier.

specifically, we have a new structure to dns subsystem:

* dns::QueryJob_Base

base type for holding a dns query and response with virtual methods
in charge of sending a reply to whoever requested.

* dns::PacketSource_Base

base type for reading and writing dns messages to and from wherever they came from

* dns::Resolver_Base

base type for filtering and handling of dns messages asynchronously.

* dns::Server

contextualized per endpoint dns object, responsible for all dns related isms.

this change hides all impelementation details of all of the dns components.
adds some more helper functions for parsing dns and dealing with OwnedBuffer.

overall dns becomes less of a pain with this new structure. probably.

Commit:
wire up dns srv records

majestrate committed:
restucture dbus parts

* move dbus into llarp/linux/dbus.hpp and llarp/linux/dbus.cpp
* provide platform abstraction for setting dns in preparation for network manager

majestrate committed:
remove compat wrapper

majestrate committed:
unbreak android config loading

majestrate committed:
changeset for windows port

* wintun vpn platform for windows
* bundle config snippets into nsis installer for exit node, keyfile persisting, reduced hops mode.
* use wintun for vpn platform
* isolate all windows platform specific code into their own compilation units and libraries
* split up internal libraries into more specific components
* rename liblokinet.a target to liblokinet-amalgum.a to elimiate ambiguity with liblokinet.so
* DNS platform for win32
* rename llarp/ev/ev_libuv.{c,h}pp to llarp/ev/libuv.{c,h}pp as the old name was idiotic
* split up net platform into win32 and posix specific compilation units
* rename lokinet_init.c to easter_eggs.cpp as that is what they are for and it does not need to be a c compilation target
* add cmake option STRIP_SYMBOLS for seperating out debug symbols for windows builds
* intercept dns traffic on all interfaces on windows using windivert and feed it into lokinet

majestrate committed:
tweaks for wine and yarn for gui

* allow specifying a custom yarn binary for building the gui using -DYARN= cmake option
* unset DISPLAY when calling wine because i hate popups
* do not rebuild gui when building for windows
* by setting the magical undocumented env var USE_SYSTEM_7ZA to 'true' we can have the pile of npm bullshit code use our system's local 7z binary instead of the probably not backdoored binary from npm, yes for real. i hate nodejs so god damn much you have no fucking idea
* allow providing a custom gui from a zip file via -DGUI_ZIP_FILE cmake option

majestrate committed:
disable lokinet-bootstrap on windows builds

jagerman committed:
Set various -W flags on apple

Apple supports anything here that Clang supports and should have them
set the same as everywhere else.

Most importantly this gives apple the -Wno-deprecated-declarations flag
which has been driving me nuts on macos.

This also version-gates the -Wno-deprecated-declarations so that it
will turn on again when we bump the version beyond .10.

majestrate committed:
revise ./contrib/format.sh

jagerman commented on cmake/macos.cmake:
```diff
@@ -150,7 +150,12 @@ if(BUILD_PACKAGE)
       COMMAND ./seticon lokinet.icns "${lokinet_installer}.dmg"
   )
-  add_dependancies(package "${lokinet_installer}.dmg")
+  add_custom_target(apple ALL COMMAND "true" DEPENDS "${lokinet_installer}.dmg")
```
Please don't.  The `package` target is reserved only when cpack is in use, which we do not use on Apple.  Using the `package` target gives us a target that works the same way whether using cpack (non-apple) or not (apple).

majestrate replied:
deoending on the package target was breaking ci on mac

jagerman commented on contrib/format.sh:
```diff
@@ -21,7 +21,7 @@ fi
 cd "$(dirname $0)/../"
 if [ "$1" = "verify" ] ; then
-    if [ $($binary --output-replacements-xml $(find jni daemon llarp include pybind | grep -E '\.([hc](pp)?|mm?)$' | grep -v '\#') | grep '</replacement>' | wc -l) -ne 0 ] ; then
+    if [ $($binary --output-replacements-xml $(find jni daemon llarp include pybind | grep -E '\.([hc](pp)?|mm)$' | grep -v '\#') | grep '</replacement>' | wc -l) -ne 0 ] ; then
```
Why don't we just let it be reformatted, by adding the `?` into the pattern 4 lines down?

majestrate replied:
it then requires an almost full reformat of the existing .m compilation units

jagerman replied:
Good.

jagerman commented on contrib/ci/drone-static-upload.sh:
```diff
@@ -50,9 +50,9 @@ elif [ -e build-docs ]; then
     cp -av build-docs/docs/mkdocs.yml build-docs/docs/markdown "$base"
     tar cJvf "$archive" "$base"
-elif [ -e build-mac ]; then
+elif [ -e build/macos ]; then
```
Please don't.  I routinely use multiple `build*` directories, including `build` for manual builds.  Nesting it under `build/` is a nuisance in multiple ways.

jagerman commented on contrib/mac.sh:
```diff
@@ -9,14 +9,11 @@
 
-if ! [ -f LICENSE ] || ! [ -d llarp ]; then
-    echo "You need to run this as ./contrib/mac.sh from the top-level lokinet project directory"
-fi
```
Why remove this check?

jagerman commented on daemon/CMakeLists.txt:
```diff
@@ -46,15 +46,27 @@ if(WITH_BOOTSTRAP)
-  if(WIN32 AND NOT MSVC_VERSION)
+  if(WIN32)
     target_sources(${exe} PRIVATE ${CMAKE_BINARY_DIR}/${exe}.rc)
+    target_compile_options(${exe} PRIVATE $<$<COMPILE_LANGUAGE:CXX>:-fpermissive>)
```
Dear christ why?

majestrate replied:
i need to remove this, artifact from old cmake hunks removed

jagerman commented on daemon/CMakeLists.txt:
```diff
@@ -46,15 +46,27 @@ if(WITH_BOOTSTRAP)
+add_library(hax_and_shims_for_cmake INTERFACE)
+if(WIN32)
+  target_link_libraries(hax_and_shims_for_cmake INTERFACE uvw oxenmq::oxenmq -lws2_32 -lshlwapi -ldbghelp -luser32 -liphlpapi -lpsapi -luserenv)
+endif()
```
This seems counterproductive; why not just link to all of these directly in the one single place that links to this interface?

majestrate replied:
link order matters here

jagerman replied:
Doesn't answer my question as to why inject a single-use INTERFACE library here.

majestrate replied:
that is exactly what this is doing.

majestrate replied:
it needs to do that. but in a very specific order

jagerman replied:
You link to this interface in exactly one place, which is going to be equivalent to just linking to the same list of items rather than the interface at the point where you are linking to the interface.

majestrate replied:
oxemq target link order was borked and i did not want to figure out how to fix it in an upstreamable way, i used this approach

jagerman replied:
"I put in a dirty hack because I couldn't be bothered to fix it properly" does not make me happy.

majestrate replied:
it also does not make me happy, when deadlines are imposed these kinds of hacks becomes the norm. this is a consistent source of our cruft across the repos, do it right or do it twice.

jagerman commented on llarp/config/ini.cpp:
```diff
@@ -17,8 +17,8 @@ namespace llarp
-      if (!f.is_open())
-        return false;
+      if (not f.is_open())
+        throw false;
```
eep!

jagerman commented on llarp/dns/multi_platform.hpp:
```diff
@@ -0,0 +1 @@
+
```
Was this supposed to be added?

jagerman commented on llarp/dns/null_platform.hpp:
```diff
@@ -0,0 +1 @@
+
```
Was this supposed to be added?

jagerman commented on contrib/ci/drone-static-upload.sh:
```diff
@@ -77,6 +80,10 @@ $mkdirs
+if [ -e "$base-debug-symbols.tar.xz" ] ; then
+    sftp -i ssh_key -b - -o StrictHostKeyChecking=off drone@oxen.rocks <<<"put $base-debug-symbols.tar.xz $upload_to"
+fi
+
```
Would be slightly cleaner (and marginally faster) to put this in a variable and stick it into the previous SFTP invocation, something like:
```
put_debug=
if [ -e "$base-debug-symbols.tar.xz" ] ; then
    put_debug="put $base-debug-symbols.tar.xz"
fi
```
...
```
sftp -i ssh_key -b - -o StrictHostKeyChecking=off drone@oxen.rocks <<SFTP
$mkdirs
put $archive $upload_to
$put_debug
SFTP
```

jagerman commented on llarp/util/compare_ptr.hpp:
```diff
@@ -16,4 +17,16 @@ namespace llarp
     }
   };
+
+  /// type for comparing weak_ptr
```
```suggestion
  /// type for comparing weak_ptr by value
```
(to disambiguate from std::owner_less, which gives comparison by owned pointer).

jagerman committed:
Make windows-configure.sh runnable without root/build

For when you want to set up a windows build dir, without doing the build
yet.

jagerman committed:
C-cast to work around gross enum==int assumption in windows API

majestrate committed:
fix up CI

add jason's suggested changes for artifact upload

use lokinet-ci-nodejs-lts as base image so we can build the installer

update ci pipeline for windows to have building gui toggle-able

by default we will build the gui from this repo, but this allows it to
easily run using a custom gui asset if needed

majestrate committed:
unit tests

jagerman committed:
Apply some lipstick to llarp_buffer_t

This class is cursed, but also broken under gcc-12.  Apply some lipstick
to get it moving again (but we really need to refactor this because it
is a mess).

jagerman committed:
Fix read problem in linux

errno is only set if read returns < 0 and won't be set to 0 if read
succeeds, so we were bailing here frequently on successful reads
(whenever errno happened to be non-0).

jagerman committed:
llarp_buffer_t: rename badly named operator==

It didn't do equality, it did "does the remaining space start with the
argument" (and so the replacement in the previous commit was broken).

This renames it to avoid the confusion and restores to what it was doing
on dev.

jagerman committed:
Fix buffer_printer overflow

majestrate committed:
small optimizations and fixes

- Ensure ip header struct is packed
- Use fmt
- add missing header

jagerman committed:
DNS: default to 127.3.2.1 & high port on Linux

Lots of tools struggle with non-default DNS port, so keep a listener on
127.3.2.1:53 (by default).

This required various changes to the config handling to hold a vector
(instead of an optional) of defaults and values, and now allows passing
in an array of defaults instead of just a single default.

jagerman committed:
Bump cmake min to 3.13 and adopt policies up to 3.24

3.13...3.xx means "minimum is 3.13, but use any new cmake policies
introduced up to 3.xx".

There was, in particular, a policy w.r.t.  external project timestamps
causing warnings under 3.24.

jagerman committed:
Avoid strict aliasing warning on function pointers

Cast via an ordinary function pointer rather than a function pointer
reference to avoid the warning.

Also make the pointer in `Func_t` explicit rather than implicit (deduced
into the `Func_t` type) to make it clearer what is going on here.

jagerman committed:
Simplify/fix ip_header layout

ip_header wasn't 20 bytes on windows compilations for some unholy
reason.  This restructures it to avoid the template and just use two
different structs for le/be with a condition_t for the ifdef, which
resolves it (and *also* apparently avoids the need for the pack).

Also add a static_assert to check the size.

Also do the same for ipv6.

jagerman committed:
Slightly DRY thread-setting code, fix warning

There were warnings from the rc variable being unused; this DRYes it to
use the same code as linux (including the failure check on rc).

jagerman committed:
Fail if wintun can't retrieve its version

majestrate committed:
promote log statement

jagerman committed:
Simplify dll loading via static function pointers

- Replaces RAII handling of DLLs with global function pointers.  (We
  don't unload the dll this way, but that seems unnecessary anyway).
- Simplifies code by just needing to call an init function, but not
  needing to pass around an object holding the function pointers.
- Adds a templated dll loader that takes the dll and a list of
  name/pointer pairs to load the dll and set the pointers in one shot.

jagerman committed:
bind/close to find free UDP port

The current code isn't working and gives a 0 (which then fails unbound
initialization).  This replaces it by doing a socket+bind to find a free
port then immediately closes (but passes the port we got into unbound).

jagerman committed:
llarp/dns logging refactor

Convert everything in llarp/dns to new-style logging.

jagerman committed:
format

jagerman committed:
Fix sockaddr_len initial value

jagerman committed:
Fixed crash in DNS resolving

The inner lambda here wasn't keeping the `Query` (`this`) alive, so
`src` wasn't valid anymore.  This changes it to copy the `src`
shared_ptr into the lambda instead of capturing `this`, and fixes it.

jagerman committed:
fix speeling

jagerman committed:
Fix backwards from/to in PacketSource_Wrapper

This resulted in DNS responses in Windows having reversed direction when
reinjected, and thus not arriving as expected.

majestrate commented on llarp/dns/server.cpp:
```diff
@@ -411,22 +411,22 @@ namespace llarp::dns
                 &Resolver::Callback,
                 nullptr))
         {
           // take back ownership on fail
```
remove comment

jagerman committed:
De-inline vpn/win32.hpp

jagerman committed:
Upgrade unbound to latest stable

jagerman committed:
Disable upstream TCP DNS

windivert, in particular, will get filtered by this and it almost
certainly won't work.

jagerman committed:
Restore Apple hax

jagerman committed:
Log upstream DNS servers when setting it up

jagerman committed:
Windows DNS fixes

- windivert was being set up *before* DNS is set up, so the DNS port was
  nullopt and thus we couldn't properly identify upstream DNS traffic.
- close() doesn't close a socket on Windows, so the socket-bind-close
  approach to get a free UDP port wasn't actually closing, and thus
  unbound upstream constrained to the given port were completely
  failing.
- The unbound thread was accessing the same shared_ptr instance as the
  outer code, which isn't thread-safe; changed it to copy a weak_ptr
  into the lambda instead.
- Exclude upstream DNS traffic in the filter rather than capturing and
  reinjecting it.

jagerman committed:
cleanups/simplifications

majestrate committed:
fix up win32 route poker a bit

majestrate committed:
limit route poker

majestrate committed:
idempotent dns

jagerman committed:
Fix crash on shutdown

For some (wrong) reason this chunk of code was here.  Removed it to stop
a crash on shutdown.

jagerman committed:
Simplifications

- a `static` is less verbose and otherwise identical to an empty
  namespace for a single declaration like this.
- operator== on two optionals already does exactly what the `is_equal`
  lambda here is doing.
- formatting

jagerman committed:
Don't do route poking when disabled

IsEnabled() is a bit broader than just IsServiceNode, so use it instead.

jagerman committed:
Linux route poker fixes

- don't add routes when not in exit mode
- don't call Up() from rpc code (the RouterPoker itself decides when to
  call Up())

jagerman committed:
Add missing != operator to nuint_t

jagerman committed:
Refactor/fix GetAdaptersAddresses

- We need to pass a flag to get Windows to include gateway info.
- Refactor it to use microsoft's recommended magic default 15000 buffer
  size and repeat in a loop a few times until it works.  Developers,
  developers, developers, developers!

jagerman committed:
Fix linked list iteration for windows ip/gateways

We were requiring `->Next` be true, which means we skipped the last (and
often only) entry of the linked lists and so never properly found the
gateway.

majestrate committed:
make unit tests pass

changes to how config defaults, specifically allowing defaults to be a
vector, broke unit test compilation. this makes them compile again.

jagerman committed:
OpenSSL 1.1.1 -> 3.0.5

jagerman committed:
macos compilation fixes

majestrate committed:
remove hunk of win32 specific code, it is dead in this codepath

jagerman committed:
Fix macos

We don't have a resolver on macos, so we were running through this loop
with fails == 0 == m_Impls.size() and throwing, crashing the process.

Early return to avoid the failure and fix macos crash.

jagerman committed:
Apple DNS configuration fix: don't obliterate trampoline

On Apple, the network extension is outside the tunnel routing, so we
cannot have libunbound talk directly to upstream (it would leak DNS when
exit mode is enabled).  Instead unbound *always* talks to a localhost
port where we have a "dns trampoline" that takes UDP packets and shoves
them through the tunnel.

We were doing that already, but recent changes here were overwriting the
libunbound settings with.

This also moves the upstream DNS configuration part of `Up()` into its
own method.

jagerman committed:
Apple DNS fix WIP

majestrate committed:
rewire up dns reconfiguration for macos

majestrate committed:
override method

jagerman committed:
Fix apple dns, part 817

jagerman committed:
ReconfigureDNS fixes, fixes macos exit mode

- ReconfigureDNS wasn't returning the old servers; made it void instead
  (the Apple code can just store a copy of the original upstream
  servers instead).
- Reconfiguring DNS reset the unbound context but didn't replace it, so
  a Down()/Up() would crash.
- Simplify Resolver() destructor to just call Down(), and make it final
  just so that no one tries to inherit from us (so that calling a
  virtual function from the destructor is safe).
- Rename CancelPendingQueries() to Down(); the former cancelled but also
  shut down the object, so the name seemed a bit misleading.
- Rename SetInternalState in Resolver_Base to ResetResolver, so that we
  aren't conflicting with ResetInternalState from Endpoint (which was a
  problem because TunEndpoint inherited from both; it could be resolved
  through the different argument type if we removed the default, but
  that seems gross).
- Make Resolver use a bare unbound context pointer rather than a
  shared_ptr; since Resolver (now) entirely manages it already we don't
  need an extra management layer, and it saves a bunch of `.get()`s.

jagerman committed:
Format fixes & fix warning in format script

- Don't escape '#' in the greps in format.sh: they warn about a spurious
  `\` in the latest grep.
- reformat

jagerman committed:
compiler warnings cmake 3.13 compatibility fix

jagerman committed:
openssl 3 build fixes

SYSTEM/MACHINE apparently doesn't work anymore and you have to pass the
system-machine value into the (cursed nasty hacky perl) configure
script.

jagerman committed:
Windows compilation fix

jagerman committed:
macos CI fixes

- Split up mac.sh into a configure + build scripts (like Windows).
- Don't attempt to build the 'package' target in CI: apparently you have
  to have a logged in user at the GUI in order to build a .dmg because
  being obtuse is the Apple way.
- Upload the raw Lokinet unsigned app in a .tar.xz, rather than dmg,
  because of the above.
- make mac.sh respect JOBS (pun not intended (but still good))

jagerman committed:
Android build fixes for openssl 3.0.5

- ANDROID_NDK_ROOT must be set in env
- cmake should be setting `-DANDROID_API=23`
- specify the correct android API via a define when building openssl; it
  has to be in CPPFLAGS (not CFLAGS) because otherwise openssl's
  configure script doesn't notice and overrides our define with the
  latest API version.
- openssl configure puts $(ANDROID_NDK_ROOT) in the makefile, so we have
  to be sure that we put it in the environment for the build command,
  too.

jagerman committed:
Disable installer path modification

This is not likely to be usable to many people, and people who it *is*
useful for are knowledgeable enough to modify it themselves.  Most users
get no use at all and it most likely just confuses them instead.

jagerman committed:
Add oxen build mirror to linux static builds

jagerman committed:
Disable mips cross compile build on ci

Nickck182 commented:
Only hard work can make progress.
